experimental/dds/tree2/src/feature-libraries/object-forest/objectForest.ts

```diff
@@ -126,74 +130,88 @@ class ObjectForest extends SimpleDependee implements IEditableForest {
 				this.forest.activeVisitor = undefined;
 				this.forest.events.emit("afterChange");
 			},
-			destroy(range: Range): void {
-				this.detachEdit(range, undefined);
+			destroy(detachedField: FieldKey, count: number): void {
+				this.forest.invalidateDependents();
+				this.forest.delete(detachedField);
 			},
-			create(index: PlaceIndex, content: Delta.ProtoNodes): void {
+			create(content: Delta.ProtoNodes, destination: FieldKey): void {
 				this.forest.invalidateDependents();
-				const nodes = Array.from(content, mapTreeFromCursor);
-				const [parent, key] = this.cursor.getParent();
-				const destinationField = getMapTreeField(parent, key, true);
-				assertValidIndex(index, destinationField, true);
-				// TODO: this will fail for very large insertions due to argument limits.
-				destinationField.splice(index, 0, ...nodes);
+				this.forest.add(content, destination);
 			},
-			attach(source: DetachedRangeUpPath, destination: PlaceIndex): void {
+			attach(source: FieldKey, count: number, destination: PlaceIndex): void {
 				this.forest.invalidateDependents();
-				this.attachEdit(source, destination);
+				this.attachEdit(source, count, destination);
 			},
-			detach(source: Range, destination: DetachedPlaceUpPath): void {
+			detach(source: Range, destination: FieldKey): void {
 				this.forest.invalidateDependents();
 				this.detachEdit(source, destination);
 			},
-			attachEdit(source: DetachedRangeUpPath, destination: PlaceIndex): void {
+			/**
+			 * Attaches the range into the current field by transferring it from the given source path.
+			 * Does not invalidate dependents.
+			 * @param source - The the range to be attached.
+			 * @param destination - The index in the current field at which to attach the content.
+			 */
+			attachEdit(source: FieldKey, count: number, destination: PlaceIndex): void {
+				assertNonNegativeSafeInteger(count);
+				if (count === 0) {
+					return;
+				}
 				const [parent, key] = cursor.getParent();
+				assert(
+					parent !== this.forest.roots || key !== source,
+					"Attach source field must be different from current field",
+				);
 				const currentField = getMapTreeField(parent, key, true);
 				assertValidIndex(destination, currentField, true);
-				const sourceField = getMapTreeField(this.forest.roots, source.field, false);
-				assertValidIndex(source.start, sourceField, true);
-				assertValidIndex(source.end, sourceField, true);
-				assert(source.end >= source.start, "Range end must be >= its start");
-				const content = sourceField.splice(source.start, source.end - source.start);
-				if (sourceField.length === 0) {
-					this.forest.roots.fields.delete(source.field);
-				}
+				const sourceField = getMapTreeField(this.forest.roots, source, false);
+				assert(
```

Comment:
generally its better to do 2 assert than one with conditions anded together since you get better errors that way.